classes/api.php

```diff
@@ -0,0 +1,317 @@
+                    str_pad($counter++, 6, '0', STR_PAD_LEFT)
+                );
+
+                $request = new Request('PUT', $this->build_blob_block_url($key, $blockid), ['content-md5' => $blockmd5], $content);
+                $promises[] = $this->client->sendAsync($request)->then(null, $this->clean_exception_sas_if_needed());
+                $blockids[] = $blockid;
+            };
+
+            if (count($blockids) > self::MAX_NUMBER_BLOCKS) {
+                throw new coding_exception("Max number of blocks reached, block size too small ?");
+            }
+
+            // Will throw exception if any fail - if any fail we want to abort early.
+            Utils::unwrap($promises);
+
+            // Commit the blocks together into a single blob.
+            $body = $this->make_block_list_xml($blockids);
+            $bodymd5 = base64_encode(hex2bin(md5($body)));
+            $request = new Request('PUT', $this->build_blocklist_url($key),
+                ['Content-Type' => 'application/xml', 'content-md5' => $bodymd5], $body);
+            $this->client->sendAsync($request)->then(null, $this->clean_exception_sas_if_needed())->wait();
+
+            // Now it is combined, set the md5 and content type on the completed blob.
+            $request = new Request('PUT', $this->build_blob_properties_url($key), [
+                'x-ms-blob-content-md5' => base64_encode($md5),
+                'x-ms-blob-content-type' => $contenttype,
+            ]);
+            $this->client->sendAsync($request)->then(null, $this->clean_exception_sas_if_needed())->wait();
+
+            // Done, resolve the entire promise.
+            $entirepromise->resolve('fulfilled');
+        });
+
+        return $entirepromise;
+    }
+
+    /**
+     * If the stream should upload using multipart upload.
+     * @param StreamInterface $stream
+     * @return bool
+     */
+    private function should_stream_upload_multipart(StreamInterface $stream): bool {
+        return $stream->getSize() > self::MULTIPART_THRESHOLD;
+    }
+
+    /**
+     * Generates a blocklist XML.
+     * @see https://learn.microsoft.com/en-us/rest/api/storageservices/put-block-list#request-body
+     * @param array $blockidlist list of block ids.
+     * @return string blocklist xml string.
+     */
+    private function make_block_list_xml(array $blockidlist): string {
+        // We use 'Latest' since we don't care about committing different
+        // blob block versions - we always want the latest.
+        $string = "<?xml version=\"1.0\" encoding=\"utf-8\"?>\n<BlockList>";
+        foreach ($blockidlist as $blockid) {
+            $string .= "\n<Latest>" . $blockid . '</Latest>';
+        }
+        $string .= "\n</BlockList>";
+        return $string;
+    }
+
+    /**
+     * Returns a request exception handling function that redacts the SAS token from error messages if needed.
+     * @return callable
+     */
+    private function clean_exception_sas_if_needed(): callable {
```

Comment:
can we just call this clean_exception_sas_token? seems a bit cleaner, I dont think the if_needed is required

Reply:
i mean, it doesn't 100% clean the exception sas, only if the flag/setting is enabled. So technically is is `if_needed` - while a bit verbose I think its good as it conveys it doesn't do it 100% of the time